app/src/ui/rebase/rebase-flow.tsx

```diff
@@ -0,0 +1,444 @@
+            manualResolutions,
+          },
+        })
+      } else if (this.state.progress.value >= 1) {
+        // waiting before the CSS animation to give the progress UI a chance to
+        // show it reaches 100%
+        await timeout(1000)
+
+        this.setState(
+          {
+            step: {
+              kind: RebaseStep.Completed,
+            },
+          },
+          () => this.props.onFlowEnded()
+        )
+      }
+    }
+
+    if (
+      this.state.step.kind === RebaseStep.ShowConflicts &&
+      // skip re-running this check once any resolved files have been detected
+      !this.state.userHasResolvedConflicts
+    ) {
+      const { workingDirectory, conflictState } = this.props
+
+      if (conflictState === null) {
+        // no conflicts found, ignoring
+        return
+      }
+
+      const resolvedConflicts = getResolvedFiles(
+        workingDirectory,
+        conflictState.manualResolutions
+      )
+
+      if (resolvedConflicts.length > 0) {
+        this.setState({
+          userHasResolvedConflicts: true,
+        })
+      }
+    }
+  }
+
+  private moveToShowConflictedFileState = () => {
+    const { workingDirectory, conflictState } = this.props
+
+    if (conflictState === null) {
+      log.error('[RebaseFlow] unable to detect conflicts for this repository')
+      return
+    }
+
+    const { manualResolutions, targetBranch } = conflictState
+
+    this.setState({
+      step: {
+        kind: RebaseStep.ShowConflicts,
+        targetBranch,
+        workingDirectory,
+        manualResolutions,
+      },
+    })
+  }
+
+  private updateProgress = (progress: IRebaseProgress) => {
+    this.setState({
+      progress,
+    })
+  }
+
+  private moveToCompletedState = () => {
+    // this ensures the progress bar fills to 100%, while `componentDidUpdate`
+    // detects and handles the state transition after a period of time to ensure
+    // the UI shows _something_ before closing the dialog
+    this.setState(prevState => {
+      const { total } = prevState.progress
+      return {
+        progress: {
+          value: 1,
+          count: total,
+          total,
+        },
+      }
+    })
+  }
+
+  private onStartRebase = (
+    baseBranch: string,
+    targetBranch: string,
+    total: number
+  ) => {
+    if (this.state.step.kind !== RebaseStep.ChooseBranch) {
+      throw new Error(`Invalid step to start rebase: ${this.state.step.kind}`)
+    }
+
+    const startRebaseAction = async () => {
+      const result = await this.props.dispatcher.rebase(
+        this.props.repository,
+        baseBranch,
+        targetBranch,
+        {
+          start: 0,
+          total,
+          progressCallback: this.updateProgress,
+        }
+      )
+
+      if (result === RebaseResult.CompletedWithoutError) {
+        this.moveToCompletedState()
+      }
+    }
+
+    this.setState(() => ({
+      step: {
+        kind: RebaseStep.ShowProgress,
+        onDidMount: startRebaseAction,
```

Comment:
This pattern fell out of a need to smooth the transition into the progress dialog. The goal here was to not start the rebase action (either `git rebase` or `git rebase --continue`) until the progress dialog is visible, because the Git operation could complete in under a second, potentially switching the dialog out before the progress dialog was actually shown to the user.

Happy to try other approaches if you have ideas...